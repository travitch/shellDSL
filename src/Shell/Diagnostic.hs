{-# LANGUAGE GeneralizedNewtypeDeriving #-}
module Shell.Diagnostic (
  Diagnostics,
  Diagnostic(..),
  recordDiagnostic,
  runDiagnostics
  ) where

import Control.Applicative
import qualified Control.Monad.State.Strict as MS
import qualified Data.Foldable as F
import Data.Sequence ( Seq )
import qualified Data.Sequence as Seq

data DiagState = DiagState { diags :: !(Seq Diagnostic) }

data Location = SrcLoc
              deriving (Eq, Ord, Show)

data Diagnostic = Diagnostic { dLocation :: Maybe Location
                             , dComponent :: String
                             , dMessage :: String
                             }
                deriving (Eq, Ord, Show)

newtype Diagnostics a = Diagnostics (MS.State DiagState a)
                      deriving (Functor, Applicative, Monad, MS.MonadState DiagState)

recordDiagnostic :: Diagnostic -> Diagnostics ()
recordDiagnostic d = MS.modify' $ \s -> s { diags = diags s Seq.|> d }

runDiagnostics :: Diagnostics a -> (a, [Diagnostic])
runDiagnostics (Diagnostics action) =
  let (res, DiagState s) = MS.runState action (DiagState Seq.empty)
  in (res, F.toList s)
