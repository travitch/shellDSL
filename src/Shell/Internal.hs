{-# LANGUAGE DeriveFunctor #-}
{-# LANGUAGE FlexibleInstances #-}
{-# LANGUAGE TupleSections #-}
{-# LANGUAGE ViewPatterns #-}
module Shell.Internal (
  -- * Running things
  run,
  background,
  wait,
  setEnv,
  unsetEnv,
  exportEnv,
  whileM,
  untilM,
  subshellM,
  subshellCaptureM,
  comment,
  exit,
  -- * Strings and names
  envRef,
  unsafeEnvRef,
  capture,
  pidOf,
  exitCode,
  anyChars,
  anyChar,
  charSet,
  negCharSet,
  -- Commands
  Command(..),
  command,
  (*|*),
  (*||*),
  (*&&*),
  (#),
  (|>),
  (|>>),
  (<|),
  (@>),
  subshell,
  -- * Tests
  testFileExists,
  -- * Internal
  BWord(..),
  BSpan(..),
  Wildcard(..),
  StreamSpec(..),
  streamSpec,
  isNullStreamSpec,
  Stream(..),
  CommandSpec(..),
  Result(..),
  Async(..),
  ShellState(..),
  Shell(..),
  TestSpec(..),
  flattenShell,
  traverseShell,
  traverseShell_,
  ShellM
  ) where

import Control.Applicative
import qualified Control.Concurrent.Supply as U
import qualified Control.Monad.Free as FR
import qualified Control.Monad.State.Strict as MS
import qualified Data.Foldable as F
import Data.Monoid
import Data.Sequence ( Seq )
import qualified Data.Sequence as Seq
import Data.String
import qualified Data.Traversable as T

data ShellState =
  ShellState { sIdSrc :: !U.Supply
             }

type ShellM = MS.StateT ShellState (FR.Free ShellF)
type FreeShell = FR.Free ShellF ()

takeId :: ShellM Int
takeId = do
  s <- MS.gets sIdSrc
  let (nid, s') = U.freshId s
  MS.put $ ShellState s'
  return nid

evalBody :: ShellM a -> ShellM (FR.Free ShellF a)
evalBody body = do
  s <- MS.gets sIdSrc
  let (local, sub) = U.splitSupply s
      body' = MS.evalStateT body (ShellState sub)
  MS.put (ShellState local)
  return body'

-- Ways to reference input streams:
--
-- * Named file
--
-- * Extracted from a 'Result'
data Stream = StreamFile Int BWord
            | StreamAppend Int BWord
            | StreamFD Int Int
            | StreamInput BWord
            deriving (Eq, Ord, Show)

-- | A specification of stream mappings.
--
-- Stdin is 0, stdout is 1, and stderr is 2.  Other streams are allowable.
data StreamSpec =
  StreamSpec { ssSpecs :: Seq Stream }
  deriving (Eq, Ord, Show)

instance Monoid StreamSpec where
  mempty = StreamSpec { ssSpecs = Seq.empty }
  mappend s1 s2 = StreamSpec { ssSpecs = ssSpecs s1 <> ssSpecs s2 }

isNullStreamSpec :: StreamSpec -> Bool
isNullStreamSpec (StreamSpec s) = Seq.null s

streamSpec :: Stream -> StreamSpec
streamSpec s = StreamSpec (Seq.singleton s)

-- | Bash Words
--
-- A Bash Word is a single command name or argument.  These are mostly
-- strings mixed with variable expansions.  Strings are allowed (and
-- an appropriate IFS will be chosen).
newtype BWord = BWord { unWord :: [BSpan] }
              deriving (Eq, Ord, Show)

data BSpan = BString String
             -- ^ A plain string
           | BVariable String
             -- ^ A checked variable reference
           | BUnsafeVariable String
             -- ^ An unchecked variable reference
           | BCommand Command
             -- ^ Capture the output of a subshell
           | BExitCode Result
             -- ^ The exit code of a process
           | BPID Async
             -- ^ The PID of a process
           | BWildcard Wildcard
           deriving (Eq, Ord, Show)

instance IsString BWord where
  fromString = toBWord

data Wildcard = WQuestion
              | WAsterisk
              | WCharSet [Char]
              | WNotCharSet [Char]
              deriving (Eq, Ord, Show)

toBWord :: String -> BWord
toBWord s = BWord [BString s]

instance Monoid BWord where
  mempty = BWord []
  mappend (BWord b1) (BWord b2) = BWord (b1 <> b2)

-- | Equivalent to the @*@ glob in shell
anyChars :: BWord
anyChars = BWord [BWildcard WAsterisk]

-- | Equivalent to the @?@ glob in shell
anyChar :: BWord
anyChar = BWord [BWildcard WQuestion]

charSet :: [Char] -> BWord
charSet cs = BWord [BWildcard (WCharSet cs)]

negCharSet :: [Char] -> BWord
negCharSet cs = BWord [BWildcard (WNotCharSet cs)]

-- | Reference the PID of a background process
pidOf :: Async -> BWord
pidOf a = BWord [BPID a]

-- | Reference the exit code of a process
exitCode :: Result -> BWord
exitCode e = BWord [BExitCode e]

-- | Create a reference to an environment variable.  The script
-- generator will check to ensure that the variable MUST be defined.
-- If the variable might not be defined, script compilation will fail.
envRef :: String -> BWord
envRef s = BWord [BVariable s]

-- | Like 'env', but there is no static check to ensure the variable
-- is defined.  Use this if you are really sure, or if the environment
-- variable will be defined by a @source@.
--
-- Preferably, you could add conditional assignments (if not defined
-- ...) after your @source@ or similar call to ensure that a value
-- will be set.
unsafeEnvRef :: String -> BWord
unsafeEnvRef s = BWord [BUnsafeVariable s]

-- | Run the given command in a subshell, capturing its output as a string.
--
-- When the command is run, you cannot retrieve its exit status.  If
-- you require that, explicitly run the command with 'subshell' and
-- manually capture its output.  It is not necessary to use 'subshell'
-- and 'capture' together.
capture :: Command -> BWord
capture c = BWord [BCommand c]

data CommandSpec =
  CommandSpec { commandName :: BWord
              , commandArguments :: [BWord]
              }
  deriving (Eq, Ord, Show)

-- | Command specifications.
--
-- Any modifications to a command affect the rightmost stream
-- specifier.  Note that subshells have their own stream specifier,
-- independent of the command they execute.
--
-- FIXME: Need to unparse precedence in and/or
data Command = Command CommandSpec StreamSpec
             | And Command Command
             | Or Command Command
             | Pipe Command Command
             | Sequence Command Command
             | SubShell Command StreamSpec
             | Test TestSpec
             deriving (Eq, Ord, Show)

data TestSpec = TSFileExists BWord
              | TSFileIsRegular BWord
              | TSFileIsDirectory BWord
              | TSFileIsSymlink BWord
              | TSPipeExists BWord
              | TSFileIsReadableY BWord
              | TSFileNotEmpty BWord
              | TSFDOpenOnTerminal BWord
              | TSFileWritableY BWord
              | TSFileExecutableY BWord
              | TSFileEffectivelyOwnedY BWord
              | TSFileEffectivelyOwnedG BWord
              | TSFileNewer BWord BWord
              | TSFileOlder BWord BWord
              | TSStringEmpty BWord
              | TSStringNotEmpty BWord
              | TSStringIdentical BWord BWord
              | TSStringNotIdentical BWord BWord
              | TSStringLT BWord BWord
              | TSStringGT BWord BWord
              | TSNegate TestSpec
              | TSIntEq BWord BWord
              | TSIntNeq BWord BWord
              | TSIntLT BWord BWord
              | TSIntGT BWord BWord
              | TSIntLE BWord BWord
              | TSIntGE BWord BWord
              deriving (Eq, Ord, Show)

type UID = Int

data ShellF next = RunSyncF UID Command next
                 | RunAsyncF UID Command next
                 | WhileF UID Command FreeShell next
                 | UntilF UID Command FreeShell next
                 | IfF UID [(Command, FreeShell)] (Maybe FreeShell) next
                 | SubBlockF UID (Maybe String) FreeShell next
                   -- ^ For block-level subshells, with optional output capture
                 | WaitF UID Async next
                 | UnsetEnvF UID String next
                 | SetEnvF UID String BWord next
                 | ExportEnvF UID String next
                 | CommentF UID String next
                 | ExitF UID (Maybe BWord) next
                 | Done
                 deriving (Eq, Show, Functor)

data Shell = RunSync UID Command
           | RunAsync UID Command
           | While UID Command [Shell]
           | Until UID Command [Shell]
           | If UID [(Command, [Shell])] (Maybe [Shell])
           | SubBlock UID (Maybe String) [Shell]
             -- ^ Subshell with an optional capture of the output
           | Wait UID Async
           | UnsetEnv UID String
           | SetEnv UID String BWord
           | ExportEnv UID String
           | Comment UID String
           | Exit UID (Maybe BWord)

modifyStreamSpec :: Command -> (StreamSpec -> StreamSpec) -> Command
modifyStreamSpec c f =
  case c of
    SubShell cmds spec -> SubShell cmds (f spec)
    Command cmd spec -> Command cmd (f spec)
    And lhs rhs -> And lhs (modifyStreamSpec rhs f)
    Or lhs rhs -> Or lhs (modifyStreamSpec rhs f)
    Pipe lhs rhs -> Pipe lhs (modifyStreamSpec rhs f)
    Sequence lhs rhs -> Sequence lhs (modifyStreamSpec rhs f)
    Test _ -> c

-- | The result of a synchronous command or a wait.  From these, we
-- can extract an exit code and possibly inspect stdout or stderr.
data Result = Result Int
            deriving (Eq, Ord, Show)

-- | The result of an asynchronous command.  We can wait on this, get
-- its PID, or reference stdout or stderr
data Async = Async Int
           deriving (Eq, Ord, Show)

background :: Command -> ShellM Async
background c = do
  uid <- takeId
  let res = Async uid
  _ <- FR.liftF (RunAsyncF uid c res)
  return res

run :: Command -> ShellM Result
run c = do
  uid <- takeId
  let res = Result uid
  _ <- FR.liftF (RunSyncF uid c res)
  return res

whileM :: Command -> ShellM () -> ShellM ()
whileM c body = do
  uid <- takeId
  body' <- evalBody body
  FR.liftF (WhileF uid c body' ())

untilM :: Command -> ShellM () -> ShellM ()
untilM c body = do
  uid <- takeId
  body' <- evalBody body
  FR.liftF (UntilF uid c body' ())

setEnv :: String -> BWord -> ShellM ()
setEnv name val = do
  uid <- takeId
  FR.liftF (SetEnvF uid name val ())

exportEnv :: String -> ShellM ()
exportEnv name = do
  uid <- takeId
  FR.liftF (ExportEnvF uid name ())

-- | Run the given block in a subshell
subshellM :: ShellM () -> ShellM ()
subshellM body = do
  uid <- takeId
  body' <- evalBody body
  FR.liftF (SubBlockF uid Nothing body' ())

subshellCaptureM :: String -> ShellM () -> ShellM ()
subshellCaptureM name body = do
  uid <- takeId
  body' <- evalBody body
  FR.liftF (SubBlockF uid (Just name) body' ())

unsetEnv :: String -> ShellM ()
unsetEnv name = do
  uid <- takeId
  FR.liftF (UnsetEnvF uid name ())

-- | Introduce a comment into the script source
comment :: String -> ShellM ()
comment str = do
  uid <- takeId
  FR.liftF (CommentF uid str ())

-- | Wait on an asynchronous/backgrounded task.
--
-- The call to wait is synchronous and returns an exit code
wait :: Async -> ShellM Result
wait a = do
  uid <- takeId
  let res = Result uid
  _ <- FR.liftF (WaitF uid a res)
  return res

exit :: Maybe BWord -> ShellM ()
exit mexitcode = do
  uid <- takeId
  FR.liftF (ExitF uid mexitcode ())

command :: BWord -> [BWord] -> Command
command cmd args = Command cspec mempty
  where
    cspec = CommandSpec { commandName = cmd
                        , commandArguments = args
                        }

testFileExists :: BWord -> Command
testFileExists = Test . TSFileExists

-- | Define a pipeline.
--
-- The LHS command is implicitly made asynchronous.  The RHS command
-- "owns" the command resulting from the ||| operator is referenced by
-- the RHS command.
(*|*) :: Command -> Command -> Command
(*|*) = Pipe

(*||*) :: Command -> Command -> Command
(*||*) = Or

(*&&*) :: Command -> Command -> Command
(*&&*) = And

-- | Sequence commands
(#) :: Command -> Command -> Command
(#) = Sequence

-- | Redirect stdout, overwriting the target file
(|>) :: Command -> BWord -> Command
c |> fp =
  modifyStreamSpec c $ \s -> s { ssSpecs = ssSpecs s Seq.|> StreamFile 1 fp }

-- | Redirect stout, appending to the target file
(|>>) :: Command -> BWord -> Command
c |>> fp =
  modifyStreamSpec c $ \s -> s { ssSpecs = ssSpecs s Seq.|> StreamAppend 1 fp }

(<|) :: Command -> BWord -> Command
c <| fp =
  modifyStreamSpec c $ \s -> s { ssSpecs = ssSpecs s Seq.|> StreamFile 0 fp }

-- | Redirect the first fd to the second fd
(@>) :: Command -> (Int, Int) -> Command
c @> (src, dst) =
  modifyStreamSpec c $ \s -> s { ssSpecs = ssSpecs s Seq.|> StreamFD src dst }

-- | Wrap the given command in a subshell
subshell :: Command -> Command
subshell c = SubShell c mempty


data FlatState = FlatState { sShells :: Seq Shell }
emptyFlatState :: FlatState
emptyFlatState = FlatState { sShells = Seq.empty }

type Flatten = MS.State FlatState

flattenM :: FreeShell -> Flatten ()
flattenM = FR.iterM $ \f ->
  case f of
    RunSyncF uid cmd next -> appendShell (RunSync uid cmd) >> next
    RunAsyncF uid cmd next -> appendShell (RunAsync uid cmd) >> next
    WaitF uid a next -> appendShell (Wait uid a) >> next
    UnsetEnvF uid str next -> appendShell (UnsetEnv uid str) >> next
    SetEnvF uid str val next -> appendShell (SetEnv uid str val) >> next
    ExportEnvF uid str next -> appendShell (ExportEnv uid str) >> next
    CommentF uid str next -> appendShell (Comment uid str) >> next
    ExitF uid mid next -> appendShell (Exit uid mid) >> next
    IfF uid tests melse next -> do
      tests' <- T.mapM (\(c, b) -> (c,) <$> nestedBlock b) tests
      melse' <- T.mapM nestedBlock melse
      appendShell (If uid tests' melse')
      next
    WhileF uid cond body next -> do
      body' <- nestedBlock body
      appendShell (While uid cond body')
      next
    UntilF uid cond body next -> do
      body' <- nestedBlock body
      appendShell (Until uid cond body')
      next
    SubBlockF uid mcap body next -> do
      body' <- nestedBlock body
      appendShell (SubBlock uid mcap body')
      next
    Done -> return ()

nestedBlock :: FreeShell -> Flatten [Shell]
nestedBlock f = do
  s0 <- MS.get
  MS.put emptyFlatState
  flattenM f
  s1 <- MS.get
  MS.put s0
  return $ F.toList $ sShells s1

appendShell :: Shell -> Flatten ()
appendShell sh = MS.modify $ \s -> s { sShells = sShells s Seq.|> sh }

flattenShell :: ShellM () -> IO [Shell]
flattenShell st = do
  s <- U.newSupply
  let freeShell = MS.evalStateT st (ShellState s)
      cmds = F.toList $ sShells $ MS.execState (flattenM freeShell) emptyFlatState
  return cmds

-- | 'traverse' specialized to 'Shell', which is monomorphic and can't
-- be an instance of the actual 'Traversable' class.  Nested
-- statements (e.g., loops) are processed after the statements in
-- their bodies.
--
-- This evaluation order might not be suitable for everything.
traverseShell :: (Applicative f, Monad f) => (Shell -> f Shell) -> Shell -> f Shell
traverseShell f s =
  case s of
    While uid c body -> do
      w <- While uid c <$> T.traverse (traverseShell f) body
      f w
    Until uid c body -> do
      u <- Until uid c <$> T.traverse (traverseShell f) body
      f u
    SubBlock uid ms body -> do
      b <- SubBlock uid ms <$> T.traverse (traverseShell f) body
      f b
    If uid cases melse -> do
      cases' <- mapM (\(c, body) -> (c,) <$> T.traverse (traverseShell f) body) cases
      melse' <- T.traverse (T.traverse (traverseShell f)) melse
      f (If uid cases' melse')
    RunSync {} -> f s
    RunAsync {} -> f s
    Wait {} -> f s
    UnsetEnv {} -> f s
    SetEnv {} -> f s
    ExportEnv {} -> f s
    Comment {} -> f s
    Exit {} -> f s

traverseShell_ :: (Applicative f, Monad f) => (Shell -> f ()) -> Shell -> f ()
traverseShell_ f s0 = traverseShell f' s0 >> return ()
  where
    f' s = f s >> return s
