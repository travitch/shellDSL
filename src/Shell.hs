module Shell (
  -- * Running things
  run,
  background,
  wait,
  setEnv,
  unsetEnv,
  exportEnv,
  -- * Control flow
  whileM,
  untilM,
  subshellM,
  subshellCaptureM,
  -- * Referencing things
  envRef,
  unsafeEnvRef,
  capture,
  -- * Commands
  Command,
  command,
  (*|*),
  (*||*),
  (*&&*),
  (#),
  (|>),
  (|>>),
  (<|),
  (@>),
  subshell
  ) where

import Prelude hiding ( until )

import Shell.Internal
