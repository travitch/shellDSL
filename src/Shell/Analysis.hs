-- | Static analysis of shell scripts
--
-- The analysis is aimed at providing some static checking of common
-- errors and providing information to support code generation
module Shell.Analysis (
  Analysis(..),
  analyze,
  requiresPid,
  requiresExitCode,
  errors
  ) where

import Control.Monad ( unless )
import qualified Control.Monad.State.Strict as MS
import qualified Data.Foldable as F
import Data.IntSet ( IntSet )
import qualified Data.IntSet as IS
import Data.Sequence ( Seq )
import qualified Data.Sequence as Seq
import Data.Set ( Set )
import qualified Data.Set as S
import Text.Printf ( printf )

import Shell.Diagnostic
import Shell.Internal

data Analysis =
  Analysis { aRequiredPID :: !IntSet
           , aRequiredExitCode :: !IntSet
           , aMayDefinedEnvVars :: !(Set String)
           , aMustDefinedEnvVars :: !(Set String)
           , aErrors :: !(Seq Diagnostic)
           }

errors :: Analysis -> [Diagnostic]
errors = F.toList . aErrors

requiresPid :: Analysis -> Shell -> Bool
requiresPid a sh =
  case sh of
    RunAsync uid _ -> uid `IS.member` aRequiredPID a
    _ -> False

requiresExitCode :: Analysis -> Shell -> Bool
requiresExitCode a sh =
  case sh of
    RunSync uid _ -> uid `IS.member` aRequiredExitCode a
    Wait uid _ -> uid `IS.member` aRequiredExitCode a
    _ -> False

type M = MS.State Analysis

emptyState :: Analysis
emptyState = Analysis { aRequiredPID = IS.empty
                      , aRequiredExitCode = IS.empty
                      , aMayDefinedEnvVars = S.empty
                      , aMustDefinedEnvVars = S.empty
                      , aErrors = Seq.empty
                      }

analyze :: [Shell] -> Analysis
analyze prog = MS.execState action emptyState
  where
    action = do
      mapM_ analyzeAction prog

analyzeAction :: Shell -> M ()
analyzeAction sh =
  case sh of
    Comment {} -> return ()
    RunSync _ cmd -> analyzeCommand cmd
    RunAsync _ cmd -> analyzeCommand cmd
    Wait _ (Async aid) ->
      -- Syntactically, wait can only refer to valid values, so we
      -- just record the PID reference.
      MS.modify $ \s -> s { aRequiredPID = IS.insert aid (aRequiredPID s) }
    UnsetEnv _ vname ->
      MS.modify $ \s -> s { aMayDefinedEnvVars = S.delete vname (aMayDefinedEnvVars s)
                           , aMustDefinedEnvVars = S.delete vname (aMustDefinedEnvVars s)
                           }
    SetEnv _ vname _ ->
      MS.modify $ \s -> s { aMayDefinedEnvVars = S.insert vname (aMayDefinedEnvVars s)
                           , aMustDefinedEnvVars = S.insert vname (aMustDefinedEnvVars s)
                           }
    ExportEnv _ vname -> do
      mustDef <- MS.gets aMustDefinedEnvVars
      unless (S.member vname mustDef) $ do
        recordError $ printf "Variable `%s` may not be defined at export" vname
    While _ cmd body -> do
      s0 <- MS.get
      analyzeCommand cmd
      mapM_ analyzeAction body
      controlFlowMerge s0
    Until _ cmd body -> do
      s0 <- MS.get
      analyzeCommand cmd
      mapM_ analyzeAction body
      controlFlowMerge s0
    SubBlock _ mcap body -> do
      -- Here we have an interesting case - environment effects in the
      -- body do not propagate to the caller.
      --
      -- We should probably also clear out the local environment in
      -- the body since it cannot reference non-exported vars (can it?)
      s0 <- MS.get
      mapM_ analyzeAction body
      MS.modify $ \now -> now { aMustDefinedEnvVars = aMustDefinedEnvVars s0
                               , aMayDefinedEnvVars = aMayDefinedEnvVars s0
                               }
      F.forM_ mcap $ \capvar -> do
        MS.modify $ \now -> now { aMustDefinedEnvVars = S.insert capvar (aMustDefinedEnvVars now)
                                 , aMayDefinedEnvVars = S.insert capvar (aMayDefinedEnvVars now)
                                 }

-- | Merge flow sensitive facts after a control flow merge point.
--
-- Right now, this is just the may/must def records
controlFlowMerge :: Analysis -> M ()
controlFlowMerge s0 =
  MS.modify $ \now -> now { aMustDefinedEnvVars = S.intersection (aMustDefinedEnvVars now) (aMustDefinedEnvVars s0)
                           , aMayDefinedEnvVars = S.union (aMayDefinedEnvVars now) (aMayDefinedEnvVars s0)
                           }

analyzeCommand :: Command -> M ()
analyzeCommand cmd =
  case cmd of
    And c1 c2 -> analyzeCommand c1 >> analyzeCommand c2
    Or c1 c2 -> analyzeCommand c1 >> analyzeCommand c2
    Pipe c1 c2 -> analyzeCommand c1 >> analyzeCommand c2
    Sequence c1 c2 -> analyzeCommand c1 >> analyzeCommand c2
    SubShell c spec -> analyzeCommand c >> analyzeStreamSpec spec
    Test t -> analyzeTestSpec t
    Command cspec sspec -> analyzeCommandSpec cspec >> analyzeStreamSpec sspec

-- | If a referenced environment variable is not defined, report an error
analyzeBWord :: BWord -> M ()
analyzeBWord = mapM_ analyzeBSpan . unWord

analyzeBSpan :: BSpan -> M ()
analyzeBSpan bs =
  case bs of
    BVariable vname -> do
      s <- MS.get
      -- This diagnostic will be much better once it has location
      -- information attached
      unless (S.member vname (aMayDefinedEnvVars s)) $ do
        recordError $ printf "Variable `%s` is not definitely assigned" vname
    BExitCode (Result rid) ->
      MS.modify $ \s -> s { aRequiredExitCode = IS.insert rid (aRequiredExitCode s)
                           }
    BPID (Async aid) ->
      MS.modify $ \s -> s { aRequiredPID = IS.insert aid (aRequiredPID s) }
    _ -> return ()

analyzeCommandSpec :: CommandSpec -> M ()
analyzeCommandSpec cspec = analyzeBWord (commandName cspec) >> mapM_ analyzeBWord (commandArguments cspec)

analyzeTestSpec :: TestSpec -> M ()
analyzeTestSpec ts =
  case ts of
    TSFileExists bw -> analyzeBWord bw
    TSFileIsRegular bw -> analyzeBWord bw
    TSFileIsDirectory bw -> analyzeBWord bw
    TSFileIsSymlink bw -> analyzeBWord bw
    TSPipeExists bw -> analyzeBWord bw
    TSFileIsReadableY bw -> analyzeBWord bw
    TSFileNotEmpty bw -> analyzeBWord bw
    TSFDOpenOnTerminal bw -> analyzeBWord bw
    TSFileWritableY bw -> analyzeBWord bw
    TSFileExecutableY bw -> analyzeBWord bw
    TSFileEffectivelyOwnedY bw -> analyzeBWord bw
    TSFileEffectivelyOwnedG bw -> analyzeBWord bw
    TSFileNewer bw1 bw2 -> analyzeBWord bw1 >> analyzeBWord bw2
    TSFileOlder bw1 bw2 -> analyzeBWord bw1 >> analyzeBWord bw2
    TSStringEmpty bw -> analyzeBWord bw
    TSStringNotEmpty bw -> analyzeBWord bw
    TSStringIdentical bw1 bw2 -> analyzeBWord bw1 >> analyzeBWord bw2
    TSStringNotIdentical bw1 bw2 -> analyzeBWord bw1 >> analyzeBWord bw2
    TSStringLT bw1 bw2 -> analyzeBWord bw1 >> analyzeBWord bw2
    TSStringGT bw1 bw2 -> analyzeBWord bw1 >> analyzeBWord bw2
    TSNegate ts' -> analyzeTestSpec ts'
    TSIntEq bw1 bw2 -> analyzeBWord bw1 >> analyzeBWord bw2
    TSIntNeq bw1 bw2 -> analyzeBWord bw1 >> analyzeBWord bw2
    TSIntLT bw1 bw2 -> analyzeBWord bw1 >> analyzeBWord bw2
    TSIntGT bw1 bw2 -> analyzeBWord bw1 >> analyzeBWord bw2
    TSIntLE bw1 bw2 -> analyzeBWord bw1 >> analyzeBWord bw2
    TSIntGE bw1 bw2 -> analyzeBWord bw1 >> analyzeBWord bw2

analyzeStreamSpec :: StreamSpec -> M ()
analyzeStreamSpec (StreamSpec specs) = F.mapM_ analyzeStream specs
  where
    analyzeStream s =
      case s of
        StreamFile _ bw -> analyzeBWord bw
        StreamAppend _ bw -> analyzeBWord bw
        StreamFD _ _ -> return ()

recordError :: String -> M ()
recordError msg =
  MS.modify $ \s -> s { aErrors = aErrors s Seq.|> d }
  where
    d = Diagnostic Nothing "Analysis" msg
