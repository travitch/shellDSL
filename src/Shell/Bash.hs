module Shell.Bash ( runBash ) where

import qualified Control.Monad.State.Strict as MS
import qualified Data.ByteString.Lazy as BS
import qualified Data.ByteString.Lazy.Builder as BS
import Data.Monoid
import Text.Printf ( printf )
import qualified Text.PrettyPrint.Mainland as PP

import Shell.Internal
import Shell.Formatter.Base

bashFormatter :: Formatter
bashFormatter = defaultFormatter

runBash :: ShellM () -> IO String
runBash st = do
  shell <- flattenShell st
  return $ renderScript bashFormatter shell

data RenderState =
  RenderState { sBuilder :: !PP.Doc
              , sFormat :: Formatter
              }

emptyRenderState :: Formatter -> RenderState
emptyRenderState fmt =
  RenderState { sBuilder = mempty
              , sFormat = fmt
              }

type Render = MS.State RenderState

renderScript :: Formatter -> [Shell] -> String
renderScript fmt s = PP.pretty 0 $ sBuilder comp
  where
    comp = MS.execState (mapM_ renderScriptM s) (emptyRenderState fmt)

-- This function will have to handle the recursive traversal of block
-- structures.  The formatters can't recurse within the Render monad,
-- so we have to do it here.  Formatters will need to be flexible to
-- return intermediate strings.
renderScriptM :: Shell -> Render ()
renderScriptM s = do
  fmt <- MS.gets sFormat
  case s of
    RunSync uid _ -> writeLineUID uid (fmtAction fmt fmt s)
    RunAsync uid _ -> writeLineUID uid (fmtAction fmt fmt s)
    _ -> writeLine (fmtAction fmt fmt s)

-- FIXME: Write a withIndentation operator to add to the current
-- indentation and then subtract when the scope is exited.  This way,
-- formatters won't need to know about it

writeLine :: PP.Doc -> Render ()
writeLine doc =
  MS.modify $ \s -> s { sBuilder = sBuilder s <> doc <> PP.line }

writeLineUID :: Int -> PP.Doc -> Render ()
writeLineUID uid doc =
  MS.modify $ \s -> s { sBuilder = sBuilder s <> doc <> PP.string str <> PP.line }
  where
    str = printf " # (uid: %d)" uid
