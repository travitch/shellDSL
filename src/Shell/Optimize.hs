{-# LANGUAGE OverloadedStrings #-}
module Shell.Optimize (
  optimize,
  Optimizer(..),
  defaultOptimizer,
  (>=>*),
  -- * Pre-defined optimizations
  optAndTrue,
  optOrFalse
  ) where

import Control.Applicative

import Shell.Diagnostic
import Shell.Internal

data Optimizer =
  Optimizer { optCommand :: Optimizer -> Command -> Diagnostics Command
              -- ^ Applied to each top-level 'Command' in a 'Shell'
              -- action.  Top-level means that the optimization driver
              -- does not recursively traverse commands itself (but
              -- the optimizer function is free to do so).
            , optAction :: Optimizer -> Shell -> Diagnostics Shell
            }

defaultOptimizer :: Optimizer
defaultOptimizer = Optimizer { optCommand = optimizeCommand
                             , optAction = optimizeAction
                             }

optimize :: Optimizer -> [Shell] -> ([Shell], [Diagnostic])
optimize o prog = runDiagnostics $ mapM (optAction o o) prog

optimizeCommand :: Optimizer -> Command -> Diagnostics Command
optimizeCommand = optAndTrue >=>* optOrFalse

-- | Compose optimization functions
(>=>*) :: (Optimizer -> Command -> Diagnostics Command)
          -> (Optimizer -> Command -> Diagnostics Command)
          -> (Optimizer -> Command -> Diagnostics Command)
f1 >=>* f2 = \o c -> f1 o c >>= f2 o

-- | Turn @command && true@ into @command@
optAndTrue :: Optimizer -> Command -> Diagnostics Command
optAndTrue _ c =
  case c of
    And lhs (Command cspec sspec)
      | isConstantCommand cspec "true" && isNullStreamSpec sspec -> do
          diag "Simplifying trivial `&& true`"
          return lhs
    _ -> return c

diag :: String -> Diagnostics ()
diag = recordDiagnostic . Diagnostic Nothing "Optimizer"

-- | Turn @command || false@ into @command@
optOrFalse :: Optimizer -> Command -> Diagnostics Command
optOrFalse _ c =
  case c of
    Or lhs (Command cspec sspec)
      | isConstantCommand cspec "false" && isNullStreamSpec sspec -> do
          diag "Simplifying trivial `|| false`"
          return lhs
    _ -> return c


isConstantCommand :: CommandSpec -> BWord -> Bool
isConstantCommand cspec name = commandName cspec == name && null (commandArguments cspec)

optimizeAction :: Optimizer -> Shell -> Diagnostics Shell
optimizeAction o = traverseShell opt
  where
    optc = optCommand o o
    opt s =
      case s of
        While uid c body -> While uid <$> optc c <*> pure body
        Until uid c body -> Until uid <$> optc c <*> pure body
        RunSync uid c -> RunSync uid <$> optc c
        RunAsync uid c -> RunAsync uid <$> optc c
        _ -> pure s
