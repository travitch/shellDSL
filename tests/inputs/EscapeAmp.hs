{-# LANGUAGE OverloadedStrings #-}
module EscapeAmp ( script ) where
import Prelude
import Shell

-- We have to escape (or quote) a string literal with an embedded &

script :: ShellM ()
script = do
  run $ command "echo" ["foo&bar"]
  return ()
