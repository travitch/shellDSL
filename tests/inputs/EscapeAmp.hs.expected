Simple { expectedOutput = "foo&bar\n"
       , diagnosticCount = 0
       }